src/network/networkd-sysctl.c

```diff
@@ -5,15 +5,196 @@
 #include <linux/if_arp.h>
 
 #include "af-list.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-manager.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+#define CGROUP_MOUNT_DFLT "/sys/fs/cgroup"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int open_rootcg(void) {
+        int fd = open(CGROUP_MOUNT_DFLT, O_PATH | O_DIRECTORY | O_CLOEXEC);
+
+        if (fd < 0)
+                log_error_errno(errno, "Failed to open cgroup mount point %s: %m", CGROUP_MOUNT_DFLT);
+
+        return fd;
+}
+
+static int sysct_write_event(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        /* Ignore our own writes */
+        if (we->pid == getpid())
+                return 0;
```

Comment:
hmm, ideally we'd filter this already in kernel space? wouldn't that be easy to add?

and i think it would be nicer if we'd do this suppression filter not via pid (because of pid recycling), but via cgroup id. i.e. allow anything through that is done by networkd's cgroup id?

Reply:
To filter them in kernel space, the BPF program needs to know the PID of networkd.
That will require another map, with data to be populated.
I can pass to userspace the value of `bpf_get_current_cgroup_id()` so to filter by cgroup id.

Reply:
And filter it like:
```c
/* Ignore our own writes */
r = cg_path_get_cgroupid("/sys/fs/cgroup/system.slice/systemd-networkd.service", &cgroup_id);
if (r < 0)
        return log_warning_errno(r, "Failed to get cgroup id: %m");

log_error("write event, cgroupids: %lu, %lu", cgroup_id, we->cgroup_id);
if (we->cgroup_id == cgroup_id)
        return 0;
```

Reply:
@poettering I tried to filter the events in the kernel.
The helper to get the current task cgroup_id is `bpf_current_task_under_cgroup()` but I get this error:
```
41: (85) call bpf_current_task_under_cgroup#37
program of this type cannot use helper bpf_current_task_under_cgroup#37
processed 40 insns (limit 1000000) max_states_per_insn 0 total_states 1 peak_states 1 mark_read 1
-- END PROG LOAD LOG --
libbpf: prog 'sysctl_monitor': failed to load: -22
libbpf: failed to load object 'sysctl_monitor_bpf'
libbpf: failed to load BPF skeleton 'sysctl_monitor_bpf': -22
failed to open and load BPF object
```
For some reason this helper is not compatible to that program type.
If this will change in the future, we can switch to kernel side filtering, but at the moment I don't think it's possible.

Reply:
It *could* be possible to have it working as you want with this kernel patch: https://github.com/teknoraver/linux/commit/312f43f7413dcbccff8426a0d512d7800e9fcbf5
But I'm trying to understand if it's acceptable before sending it.

Reply:
@poettering my proposal is: filter via cgroup in systemd at the moment, and if the kernel patch is accepted, move the filter kernel side, in the eBPF.